Delete Dialog

porterehunley opened the pull request:
<img width="1440" alt="Screen Shot 2021-02-23 at 2 59 27 PM" src="https://user-images.githubusercontent.com/42723299/108900545-bafd7500-75e7-11eb-91a9-b376d5c4a7a1.png">
<img width="1440" alt="Screen Shot 2021-02-23 at 2 59 35 PM" src="https://user-images.githubusercontent.com/42723299/108900562-bfc22900-75e7-11eb-84e1-49a0145b486a.png">

The modal is shared as an injectable. Colors are the standard 'warn' colors from material

Commit:
init

Commit:
general modal and more

Commit:
deleting accounts

Commit:
styled

Commit:
linter

Commit:
merge diff

Commit:
mispelled action

grarer commented on backend/src/routers/accountsRouter.ts:
```diff
@@ -65,6 +65,16 @@ accountsRouter.get('/info', handleAsync(async (req, res) => { // TODO make this
 }));
 
+accountsRouter.delete('/delete', handleAsync(async (req, res) => {
+  const uuid = await userUuidFromReq(req);
```
this is fine, but I should note you could also use `authenticateUUID` for this purpose, since `authenticateUUID` automatically throws a 403 if the token is invalid.

grarer commented on frontend/src/app/components/common/dialog/dialog.service.ts:
```diff
@@ -0,0 +1,34 @@
+      }
+    });
+  }
+  public confirmed(): Observable<any> | undefined {
```
mixing observables and promises can be problematic, and in general promises are better because they support async/await. i'd reccomend calling `toPromise` on the observable and returning it as a promise so that consumers can await.

grarer commented on frontend/src/app/components/settings/settings.component.ts:
```diff
@@ -117,6 +121,35 @@ export class SettingsComponent implements OnInit {
+      confirmText: 'CONFIRM',
+    };
+
+    this.dialogService.open(data.title, data.body, data.cancelText, data.confirmText);
```
if you find it more conveient to construct the parameters as an object (which imo is very reasonable), then it might also be a good idea to have `dialogService` take that one object as its only parameter rather than as separate strings. this might also be helpful at making sure that callers never provide the strings in the wrong order.

porterehunley replied:
js doesn't need key word args! /s

grarer reviewed (approved):
i left some feedback on things that could be changed, but overall this looks good. i do think we should definitely switch the the dialog to using a promise rather than an observable, but you or I can make that change later if you'd rather get #25 done first, so feel free to squash-merge now if you'd like.

Commit:
req changes

GRarer commented:
something about the way that the `observable<any>` is working there seems kinda weird,and im not certain that it won't cause bugs. i'd like to do a bit more testing before we merge this

grarer committed:
refactor dialog service to use async/await

makes the dialog service more type-safe (removing the use of `any`),
and introduces async/await so that opening and waiting on the dialog
can be done in one line and without the pyramid of doom

GRarer commented:
i was able to remove the unsafe `any`s and get rid of the nested `.try`s. I also noticed that this service would misbehave if we somehow had two dialogs open at once; for now I just have it throw an error (compatible with `apiService.showErrorPopup`) if we try to open a dialog when one is already open, but if we ever actually need to we can probably come up with a way to get the dependency injector to give us two separate instances.
